Fleet order adjustments part 1

Dilvish-fo opened the pull request:
This PR includes some general cleanup of AI treatment of order issuance and execution tracking, along with a number of changes directed at improving AI exploration and a bit for AI repair.

I am thinking it might be helpful to restructure some of the existing order types to subclass from this new OrderPause, and also plan some other changes to hopefully help facilitate maneuvers like a military fleet Pounce or somesuch, but that is still rather nascent and would follow in a separate PR.  As it stands I think this PR makes some minor improvements, and at least my testing has not indicated it causing any trouble.

Dilvish-fo committed:
Adjust AI early exploration

During first 20 turns, was skipping over systems with monsterThreat > 200, now
having AI skip if there is any monster threat at all, better to save those for later

Dilvish-fo committed:
Add a current system ID getter for AI proxy fleet class

Dilvish-fo committed:
Clean up AI handling of order_issued and executed

Previously, at least some parts of the code treated these as if execution was more sure than issuance, as if their meaning was swapped, but the handling did not even seem to be consistent.

Dilvish-fo committed:
Extract AI fleetUtils function to get a fleet's current and max structure

Dilvish-fo committed:
Add some AI type hinting

Dilvish-fo committed:
Have superceding versions of AI fleet_order.issue_order likewise return bools

 - make the return type consistent with the base version to better support
   further subclassing

Dilvish-fo committed:
Have AI OrderRepair require full fleet health for completion

 - previously, had only required arrival at repair destination

Dilvish-fo committed:
Add new AI fleet order OrderPause

requires that the fleet be actually located in target system for the order execution
to be considered completed

Dilvish-fo committed:
Use the new AI fleet order OrderPause for exploration

When OrderMove was previously designated as the order type for exploration,
the AI fleet movement process forced them to stop at every system, and so a
simple MoveOrder was sufficient to cause an AI exploration fleet to visit,
and stop at, its target system.  When that move process was changed to no longer
always force stops at every system, the AI scouts would often be deemed to have
finished exploring a given system while actually still on their way to it,
and would get a new destination, causing them to merely pass through the
original target system without pausing.  This fixes that.

Dilvish-fo committed:
Tell AI savegame codec that new OrderPause is trusted

Dilvish-fo committed:
grooming

Morlic-fo reviewed (approved):
Looks like a good improvement and clears things up that were previously rather weird re the usage of executed/issued in fleet orders.

Dilvish-fo committed:
Have AI doublecheck ship fleet assignment

 - if the given ship had been transferred to a different fleet while out of sight,
and the original fleet has not been observed since then, but the ship has been observed in a new fleet, then the AI's copy of the original fleet would still list that ship as part of it, but the AI's copy of the ship would know it was part of a new fleet.

Dilvish-fo committed:
Fix a comment typo

Dilvish-fo committed:
Have AI use warn-level logging for certain order failures

Dilvish-fo committed:
Remove superfluous line
